test/src/main/java/org/axonframework/test/matchers/DeepEqualsMatcher.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2010-2022. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.test.matchers;
+
+import org.hamcrest.BaseMatcher;
+import org.hamcrest.Description;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Field;
+import java.util.Objects;
+
+import static org.axonframework.common.BuilderUtils.assertNonNull;
+import static org.axonframework.common.ReflectionUtils.hasEqualsMethod;
+
+/**
+ * A {@link BaseMatcher} implementation that first matches based on {@link Object#equals(Object)}. When this fails and
+ * {@code equals()} is not implemented by {@code T}, the fields are matched for equality. If this fails due to
```

Comment:
strictly speaking, `equals` is always implemented. Perhaps we should say " if `equals` is not overridden from `Object`, the instance's field values are checked for equality`. If this fails....."